File: packages/ai-setup-ts/src/__tests__/opencode-validate.test.ts
import { chmodSync, mkdirSync, mkdtempSync, rmSync, writeFileSync } from 'node:fs'
import { tmpdir } from 'node:os'
import path from 'node:path'
import { afterEach, beforeEach, describe, expect, it } from 'vitest'
import { type CmdRunner, validateOpenCodeInstall } from '../adapters/opencode-validate.js'

describe('validateOpenCodeInstall', () => {
  let tempDirs: string[]
  let originalPath: string | undefined

  let originalSkip: string | undefined

  beforeEach(() => {
    tempDirs = []
    originalPath = process.env.PATH
    originalSkip = process.env.AI_SETUP_SKIP_VALIDATION
    delete process.env.AI_SETUP_SKIP_VALIDATION
  })

  afterEach(() => {
    for (const dir of tempDirs) {
      rmSync(dir, { recursive: true, force: true })
    }
    if (originalPath === undefined) {
      delete process.env.PATH
    } else {
      process.env.PATH = originalPath
    }
    if (originalSkip === undefined) {
      delete process.env.AI_SETUP_SKIP_VALIDATION
    } else {
      process.env.AI_SETUP_SKIP_VALIDATION = originalSkip
    }
  })

  const makeTempDir = (prefix: string): string => {
    const dir = mkdtempSync(path.join(tmpdir(), prefix))
    tempDirs.push(dir)
    return dir
  }

  const stubOpenCodeOnPath = (): void => {
    const binDir = makeTempDir('ai-setup-validate-bin-')
    const fakeBin = path.join(binDir, 'opencode')
    writeFileSync(fakeBin, '#!/bin/sh\nexit 0\n')
    chmodSync(fakeBin, 0o755)
    process.env.PATH = binDir
  }

  it('returns no warnings when the opencode binary is absent', () => {
    process.env.PATH = ''
    const ocDir = makeTempDir('ai-setup-validate-ocdir-')

    const stubRunner: CmdRunner = () => {
      throw new Error('unexpected call — binary should be absent')
    }

    const warnings = validateOpenCodeInstall({ ocDir }, stubRunner)
    expect(warnings).toEqual([])
  })

  it('emits a config warning when opencode debug config fails', () => {
    stubOpenCodeOnPath()
    const ocDir = makeTempDir('ai-setup-validate-ocdir-')

    const stubRunner: CmdRunner = (_cmd, args) => {
      if (args[0] === 'debug' && args[1] === 'config') {
        throw new Error('config parse error')
      }
      return 'ok'
    }

    const warnings = validateOpenCodeInstall({ ocDir }, stubRunner)
    expect(warnings.some((w) => w.scope === 'config')).toBe(true)
  })

  it('emits an agent warning when opencode debug agent fails for an installed agent', () => {
    stubOpenCodeOnPath()
    const ocDir = makeTempDir('ai-setup-validate-ocdir-')
    const agentsDir = path.join(ocDir, 'agents')
    mkdirSync(agentsDir, { recursive: true })
    writeFileSync(path.join(agentsDir, 'builder.md'), '# Builder')

    const stubRunner: CmdRunner = (_cmd, args) => {
      if (args[0] === 'debug' && args[1] === 'agent') {
        throw new Error('parse error in frontmatter')
      }
      return '{"mcp": {}}'
    }

    const warnings = validateOpenCodeInstall({ ocDir }, stubRunner)
    expect(warnings.some((w) => w.scope === 'agent' && w.item === 'builder')).toBe(true)
  })

  it('returns no warnings when config and all agents pass', () => {
    stubOpenCodeOnPath()
    const ocDir = makeTempDir('ai-setup-validate-ocdir-')
    const agentsDir = path.join(ocDir, 'agents')
    mkdirSync(agentsDir, { recursive: true })
    writeFileSync(path.join(agentsDir, 'builder.md'), '# Builder')

    const stubRunner: CmdRunner = () => '{"mcp": {"servers": {}}}'

    const warnings = validateOpenCodeInstall({ ocDir }, stubRunner)
    expect(warnings).toEqual([])
  })
})

File: packages/cli/internal/adapter/opencode_validate_test.go
package adapter

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"testing"

	"github.com/rluisb/lazyai/packages/cli/internal/types"
)

func makeValidateCtx(t *testing.T) *AdapterContext {
	t.Helper()
	return &AdapterContext{
		TargetDir:  t.TempDir(),
		SetupScope: types.SetupScopeProject,
		LibraryFS:  createTestFS(),
		Strategy:   types.ConflictStrategyAlign,
	}
}

func makeFakeOpenCodeBin(t *testing.T) string {
	t.Helper()
	binDir := t.TempDir()
	fakeBin := filepath.Join(binDir, "opencode")
	content := []byte("#!/bin/sh\nexit 0\n")
	if runtime.GOOS == "windows" {
		fakeBin = filepath.Join(binDir, "opencode.bat")
		content = []byte("@echo off\r\nexit /b 0\r\n")
	}
	if err := os.WriteFile(fakeBin, content, 0o755); err != nil {
		t.Fatalf("writing fake binary: %v", err)
	}
	return binDir
}

func TestValidateOpenCodeInstall_BinaryAbsent(t *testing.T) {
	ctx := makeValidateCtx(t)

	// Stub runner that always fails — but since LookPath will fail for a
	// nonexistent binary we never reach the runner.
	stubRunner := func(name string, args ...string) ([]byte, error) {
		return nil, fmt.Errorf("unexpected call")
	}

	// Temporarily override PATH to guarantee the binary is absent.
	t.Setenv("PATH", "")

	warnings, err := validateOpenCodeInstall(ctx, stubRunner)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(warnings) != 0 {
		t.Errorf("expected no warnings when binary absent, got %v", warnings)
	}
}

func TestValidateOpenCodeInstall_ConfigFails(t *testing.T) {
	ctx := makeValidateCtx(t)

	t.Setenv("PATH", makeFakeOpenCodeBin(t))

	callCount := 0
	stubRunner := func(name string, args ...string) ([]byte, error) {
		callCount++
		if len(args) >= 2 && args[0] == "debug" && args[1] == "config" {
			return nil, fmt.Errorf("config parse error")
		}
		return []byte("ok"), nil
	}

	warnings, err := validateOpenCodeInstall(ctx, stubRunner)
	if err != nil {
		t.Fatalf("unexpected hard error: %v", err)
	}

	found := false
	for _, w := range warnings {
		if w.Scope == "config" {
			found = true
		}
	}
	if !found {
		t.Errorf("expected a config warning; got %v", warnings)
	}
}

func TestValidateOpenCodeInstall_AgentFails(t *testing.T) {
	ctx := makeValidateCtx(t)

	// Write a fake agent file so the glob finds it.
	ocDir := filepath.Join(ctx.TargetDir, ".opencode")
	agentsDir := filepath.Join(ocDir, "agents")
	if err := os.MkdirAll(agentsDir, 0o755); err != nil {
		t.Fatalf("mkdir: %v", err)
	}
	if err := os.WriteFile(filepath.Join(agentsDir, "implementer.md"), []byte("# Implementer"), 0o644); err != nil {
		t.Fatalf("write agent: %v", err)
	}

	t.Setenv("PATH", makeFakeOpenCodeBin(t))

	stubRunner := func(name string, args ...string) ([]byte, error) {
		if len(args) >= 3 && args[0] == "debug" && args[1] == "agent" {
			return nil, fmt.Errorf("parse error in frontmatter")
		}
		// config passes but returns empty (no "mcp" substring)
		return []byte(`{"mcp": {}}`), nil
	}

	warnings, err := validateOpenCodeInstall(ctx, stubRunner)
	if err != nil {
		t.Fatalf("unexpected hard error: %v", err)
	}

	found := false
	for _, w := range warnings {
		if w.Scope == "agent" && w.Item == "implementer" {
			found = true
		}
	}
	if !found {
		t.Errorf("expected agent warning for implementer; got %v", warnings)
	}
}

func TestValidateOpenCodeInstall_AllPass(t *testing.T) {
	ctx := makeValidateCtx(t)

	ocDir := filepath.Join(ctx.TargetDir, ".opencode")
	agentsDir := filepath.Join(ocDir, "agents")
	if err := os.MkdirAll(agentsDir, 0o755); err != nil {
		t.Fatalf("mkdir: %v", err)
	}
	if err := os.WriteFile(filepath.Join(agentsDir, "implementer.md"), []byte("# Implementer"), 0o644); err != nil {
		t.Fatalf("write agent: %v", err)
	}

	t.Setenv("PATH", makeFakeOpenCodeBin(t))

	stubRunner := func(name string, args ...string) ([]byte, error) {
		return []byte(`{"mcp": {"servers": {}}}`), nil
	}

	warnings, err := validateOpenCodeInstall(ctx, stubRunner)
	if err != nil {
		t.Fatalf("unexpected hard error: %v", err)
	}
	if len(warnings) != 0 {
		t.Errorf("expected no warnings on all-pass, got %v", warnings)
	}
}
